### PR TITLE
Custom RKE1 Node Cleanup

### PR DESCRIPTION
**Pull request by luthermonson**
Problem: When creating a custom RKE1 cluster the expectation was to take a fresh server with docker, copy/paste a docker run command and add a new node to a cluster. However, if the node was removed from the cluster via the Rancher UI a node remove call is made to the kubernetes api but we would leave RKE1 tooling around on the node and mangled networking/firewall configs. This led to a lot of work for support to create cleanup scripts to undo the mess. 

Solution: Upon v3 Node delete in rancher we will pause the v1 node delete until we leave a cleanup process on the deleting node. We will target a k8s job to the node with the docker socket mounted who then starts a cattle-node-cleanup container. This container is outside of k8s and won't be reaped when the v1 node is deleted and kubelet is told to shutdown. The new cleanup process will pause and wait for kubelet to remove all k8s_ containers and when that's completed it will proceed to wipe everything from the server (containers, directories, networking, firewall rules, etc). This now means you can add a node with the docker command and delete it through the UI and immediately add it back to the cluster with no further intervention.

There are also utility methods now build into the agent to allow you to get a cleanup script from the agent container and run it yourself by hand this way we can document the cleanup process using only the agent image. This solution works on both custom linux and windows nodes.

Additional: This PR addresses the old docker cli version we had in the agent, we used it for the share-mnt to patch ephemeral /etc directories in things like RancherOS. This removes the docker cli and makes docker go client calls instead within the agent.

https://github.com/rancher/rancher/issues/26545
https://github.com/rancher/rancher/issues/32304
https://github.com/rancher/rancher/issues/32052

### CODE REVIEW

File: pkg/agent/clean/node_notwindows.go [hunk 1]
```diff
@@ -0,0 +1,590 @@
...
+const (
+	DockerPipe = "/var/run/docker.sock"
+	HostMount  = "/host"
+)
+
+func Run(ctx context.Context, args []string) error {
+	if len(args) > 3 {
+		fmt.Println(usage())
+		return nil
+	}
```
Comment (aiyengar2) [inline pkg/agent/clean/node_notwindows.go:34]: Any reason why we aren't returning an error?

Comment (aiyengar2) [reply pkg/agent/clean/node_notwindows.go:34]: also, if `len(args) < 3`, why is `Node(ctx)` (cleaning?) the default behavior? Seems like a weird thing to do if the args are `agent job`.

File: pkg/controllers/management/node/cleanup.go [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+			return nil // gave it 30s and it timed out so move to delete v1 node
+		}
+		return err
+	}
+
+	// remove the job to clean up
+	return userContext.K8sClient.BatchV1().Jobs(job.Namespace).Delete(m.ctx, job.Name, metav1.DeleteOptions{})
+}
+
+func (m *Lifecycle) createCleanUpJob(userContext *config.UserContext, node *v3.Node) (*batchV1.Job, error) {
```
Comment (aiyengar2) [inline pkg/controllers/management/node/cleanup.go:193]: nit: cleanup is one word.
```suggestion
func (m *Lifecycle) createCleanupJob(userContext *config.UserContext, node *v3.Node) (*batchV1.Job, error) {
```

**Review by aiyengar2 — CHANGES_REQUESTED** [review]
Mostly looks good! Just a couple of minor comments.

File: pkg/agent/clean/node_notwindows.go (hunk 1 above)
Comment (luthermonson) [reply pkg/agent/clean/node_notwindows.go:34]: cause it prints the usage and assumes you dont know what youre doing? possible usage should be the default behavior... let me thing about this.

File: pkg/controllers/management/node/cleanup.go (hunk 2 above)
Comment (luthermonson) [reply pkg/controllers/management/node/cleanup.go:193]: how does one give dagger eyes on this thing?

**Review by aiyengar2 — DISMISSED** [review]
LGTM

**Review by phillipsj — DISMISSED** [review]

**Review by aiyengar2 — APPROVED** [review]
LGTM

File: pkg/rkeworker/execute.go [hunk 3]
```diff
@@ -29,9 +30,11 @@ func ExecutePlan(ctx context.Context, nodeConfig *NodeConfig, writeCertOnly bool
 
 	f := fileWriter{}
 	for _, file := range nodeConfig.Files {
+		logrus.Infof("writing file %s", file.Name)
 		f.write(file.Name, file.Contents)
 	}
 	if writeCertOnly {
+		logrus.Info("writing certificates only, no need to continue executing plan")
```
Comment (dramich) [inline pkg/rkeworker/execute.go:37]: Can we move these to debug?

File: pkg/agent/clean/node_notwindows.go [hunk 4]
```diff
@@ -0,0 +1,591 @@
...
+	node - cleans the entire node and performs all actions below and is the default command, requires docker socket to clean up dockerd
+	links - clean the interface links created on the host, requires --privileged and --network host
+	paths - clean the paths added by the host like /etc/rancher, /var/lib/rancher, etc. requires -v /:/host 
+	firewall - clean the iptables firewall rules, needs --privileged and --network host to work 
+	script - prints a bash script you can run to clean up the node from the cli
+	help - print this help message
+
+other commands for automation:
+	job - used by the k8s batch job to start the ` + NodeCleanupContainerName + ` container to watch kubelet cleanup and wait to clean the node
+`
```
Comment (dramich) [inline pkg/agent/clean/node_notwindows.go:415]: These will all print out indented since there is a tab at the beginning.

**Review by dramich — APPROVED** [review]
Couple small comments, looking forward to having some clean nodes

**Review by dramich — APPROVED** [review]

File: pkg/agent/clean/node_notwindows.go (hunk 4 above)
Comment (luthermonson) [reply pkg/agent/clean/node_notwindows.go]: good point.

**Review by dramich — APPROVED** [review]

**Review by dramich — APPROVED** [review]

**Comment by kinarashah** [comment]
https://github.com/rancher/rancher/pull/32899

**Pull request merged**